File: frontend/lib/api/public-data.ts
/**
 * 공공데이터포털 API 클라이언트
 * 백엔드를 통한 VWorld API 호출
 */
import apiClient from './client';
import type { StandardResponse } from '@/types/api';

export interface ProvinceItem {
  code: string;
  name: string;
}

export interface CityItem {
  code: string;
  name: string;
}

/**
 * 광역시도 목록 조회
 */
export const getProvinces = async (): Promise<ProvinceItem[]> => {
  try {
    const response = await apiClient.get<StandardResponse<ProvinceItem[]>>('/public-data/provinces');
    if (!response.data.success || !response.data.data) {
      throw new Error(response.data.error || '광역시도 목록을 불러올 수 없습니다');
    }
    return response.data.data;
  } catch (error: any) {
    console.error('광역시도 목록 API 호출 실패:', error);
    throw error instanceof Error ? error : new Error('광역시도 목록을 불러올 수 없습니다');
  }
};

/**
 * 시군구 목록 조회
 * 
 * @param provinceCode 광역시도 코드 (11, 21, 22 등)
 */
export const getCitiesByProvince = async (provinceCode: string): Promise<CityItem[]> => {
  try {
    const response = await apiClient.get<StandardResponse<CityItem[]>>('/public-data/cities', {
      params: { province_code: provinceCode },
    });
    if (!response.data.success || !response.data.data) {
      throw new Error(response.data.error || '시군구 목록을 불러올 수 없습니다');
    }
    return response.data.data;
  } catch (error: any) {
    console.error('시군구 목록 API 호출 실패:', error);
    throw error instanceof Error ? error : new Error('시군구 목록을 불러올 수 없습니다');
  }
};

File: backend/app/api/v1/public_data.py
"""
공공데이터포털 API 프록시 엔드포인트
VWorld API를 통한 행정구역 정보 조회
"""
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Dict, Any

from app.core.database import get_db
from app.schemas.vehicle import StandardResponse
from app.services.public_data_service import PublicDataService

router = APIRouter(prefix="/public-data", tags=["공공데이터"])


@router.get("/provinces", response_model=StandardResponse)
async def get_provinces():
    """
    광역시도 목록 조회 API
    
    고정된 광역시도 목록을 반환합니다.
    """
    try:
        provinces = await PublicDataService.get_all_provinces()
        return StandardResponse(
            success=True,
            data=provinces,
            error=None
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"광역시도 목록 조회 중 오류가 발생했습니다: {str(e)}"
        )


@router.get("/cities", response_model=StandardResponse)
async def get_cities(
    province_code: str = Query(..., description="광역시도 코드 (11, 21, 22 등)")
):
    """
    시군구 목록 조회 API
    
    광역시도 코드로 해당 지역의 시군구 목록을 조회합니다.
    VWorld API를 통해 조회합니다.
    """
    try:
        cities = await PublicDataService.get_cities_by_province(province_code)
        return StandardResponse(
            success=True,
            data=cities,
            error=None
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"시군구 목록 조회 중 오류가 발생했습니다: {str(e)}"
        )

